src/functional/ffront/decorator.py

```diff
@@ -167,62 +166,50 @@ def from_function(
         externals: Optional[dict] = None,
         backend: Optional[FencilExecutor] = None,
         grid_type: Optional[GridType] = None,
-    ) -> "Program":
-        captured_vars = CapturedVars.from_function(definition)
-        past_node = ProgramParser.apply_to_function(definition)
+    ) -> Program:
+        source_def = SourceDefinition.from_function(definition)
+        external_vars = {**(externals or {}), **get_external_vars(definition)}
+        annotations = typing.get_type_hints(definition)
+        past_node = ProgramParser.apply(source_def, external_vars, annotations)
         return cls(
             past_node=past_node,
-            captured_vars=captured_vars,
-            externals={} if externals is None else externals,
+            external_vars=external_vars,
             backend=backend,
             definition=definition,
             grid_type=grid_type,
         )
 
     def __post_init__(self):
-        # validate contents of captured vars
-        for name, value in self._filter_capture_vars_by_type(GTCallable).items():
-            if value.__gt_itir__().id != name:
-                raise RuntimeError(
-                    "Name of the closure reference and the function it holds do not match."
-                )
+        external_functions = _filter_external_vars_by_type(self.external_vars, GTCallable)
+        if not all(map(lambda var: var[0] == var[1].__gt_itir__().id, external_functions.items())):
+            raise RuntimeError("Symbol name and external function's name must match.")
 
-        # validate Symbols of captured vars in PAST
-        referenced_var_names: set[str] = set()
-        for captured_var in self.past_node.captured_vars:
-            if isinstance(captured_var.type, (ct.CallableType, ct.OffsetType, ct.DimensionType)):
-                referenced_var_names.add(captured_var.id)
-            else:
-                raise NotImplementedError("Only function closure vars are allowed currently.")
-        defined_var_names = set(self.all_capture_vars.globals) | set(
-            self.all_capture_vars.nonlocals
-        )
-        if undefined := referenced_var_names - defined_var_names:
-            raise RuntimeError(f"Reference to undefined symbol(s) `{', '.join(undefined)}`.")
+        for symbol in self.past_node.external_symbols:
+            if symbol.id not in self.external_vars:
+                raise RuntimeError(f"Undefined symbol: {symbol.id}")
```

Comment:
The previous version threw an error that showed you all symbols that were undefined, while now you only get an error for the first one. Additionally the previous error message was more specific. I would suggest to restore that behavior.